File: otto-android/app/src/main/java/to/ottomot/driftd/core/chat/ChatUnreadTracker.kt
package to.ottomot.driftd.core.chat

import java.time.Instant
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import to.ottomot.driftd.core.network.dto.CircleChatMessageDto
import to.ottomot.driftd.core.network.dto.DirectConversationDto
import to.ottomot.driftd.core.network.dto.DirectMessageDto

class ChatUnreadTracker(
    private val cursorStore: ChatReadCursorStore,
) {
    private val _unreadCountByCircleId = MutableStateFlow<Map<String, Int>>(emptyMap())
    val unreadCountByCircleId: StateFlow<Map<String, Int>> = _unreadCountByCircleId.asStateFlow()

    private val _unreadCountByConversationId = MutableStateFlow<Map<String, Int>>(emptyMap())
    val unreadCountByConversationId: StateFlow<Map<String, Int>> = _unreadCountByConversationId.asStateFlow()

    val totalChatUnreadCount: Int
        get() = _unreadCountByCircleId.value.values.sum() + _unreadCountByConversationId.value.values.sum()

    private var currentUserId: String = ""
    var squadChatTabVisibleCircleId: String? = null
    var directThreadVisibleConversationId: String? = null

    fun bind(currentUserIdRaw: String) {
        val trimmed = currentUserIdRaw.trim()
        if (trimmed == currentUserId) return
        currentUserId = trimmed
        cursorStore.bind(if (trimmed.isEmpty()) null else trimmed)
    }

    fun clearAll() {
        _unreadCountByCircleId.value = emptyMap()
        _unreadCountByConversationId.value = emptyMap()
        cursorStore.clearAll()
    }

    fun setSquadChatTabVisible(circleId: String?, visible: Boolean) {
        val key = circleId?.trim()?.takeIf { it.isNotEmpty() }
        squadChatTabVisibleCircleId =
            when {
                visible && key != null -> key
                !visible && (squadChatTabVisibleCircleId == key || key == null) -> null
                else -> squadChatTabVisibleCircleId
            }
    }

    fun setDirectThreadVisible(conversationId: String?, visible: Boolean) {
        val key = conversationId?.trim()?.takeIf { it.isNotEmpty() }
        directThreadVisibleConversationId =
            when {
                visible && key != null -> key
                !visible && (directThreadVisibleConversationId == key || key == null) -> null
                else -> directThreadVisibleConversationId
            }
    }

    fun markSquadReadIfChatTabVisible(circleId: String, messages: List<CircleChatMessageDto>) {
        val key = circleId.trim().takeIf { it.isNotEmpty() } ?: return
        if (squadChatTabVisibleCircleId != key) return
        val newest = messages.lastOrNull()
        if (newest == null) {
            _unreadCountByCircleId.update { it - key }
            return
        }
        cursorStore.markRead(
            thread = ChatReadThreadKey(ChatReadThreadKind.Squad, key),
            messageId = newest.id,
            readAt = ChatUnreadCounting.messageCreatedAt(newest) ?: Instant.EPOCH,
        )
        _unreadCountByCircleId.update { it - key }
    }

    fun markDirectReadIfThreadVisible(conversationId: String, messages: List<DirectMessageDto>) {
        val key = conversationId.trim().takeIf { it.isNotEmpty() } ?: return
        if (directThreadVisibleConversationId != key) return
        val newest = messages.lastOrNull()
        if (newest == null) {
            _unreadCountByConversationId.update { it - key }
            return
        }
        cursorStore.markRead(
            thread = ChatReadThreadKey(ChatReadThreadKind.Direct, key),
            messageId = newest.id,
            readAt = ChatUnreadCounting.messageCreatedAt(newest) ?: Instant.EPOCH,
        )
        _unreadCountByConversationId.update { it - key }
    }

    fun recomputeSquad(circleId: String, messages: List<CircleChatMessageDto>) {
        val key = circleId.trim().takeIf { it.isNotEmpty() } ?: return
        val count =
            ChatUnreadCounting.squadUnreadCount(
                messages = messages,
                cursor = cursorStore.cursor(ChatReadThreadKey(ChatReadThreadKind.Squad, key)),
                currentUserId = currentUserId,
                trackingStartedAt = cursorStore.trackingStartedAt,
            )
        _unreadCountByCircleId.update { map ->
            if (count > 0) map + (key to count) else map - key
        }
    }

    fun recomputeDirect(conversationId: String, messages: List<DirectMessageDto>) {
        val key = conversationId.trim().takeIf { it.isNotEmpty() } ?: return
        val count =
            ChatUnreadCounting.directUnreadCount(
                messages = messages,
                cursor = cursorStore.cursor(ChatReadThreadKey(ChatReadThreadKind.Direct, key)),
                currentUserId = currentUserId,
                trackingStartedAt = cursorStore.trackingStartedAt,
            )
        _unreadCountByConversationId.update { map ->
            if (count > 0) map + (key to count) else map - key
        }
    }

    fun recomputeDirectFromPreview(conversation: DirectConversationDto) {
        val key = conversation.id.trim().takeIf { it.isNotEmpty() } ?: return
        val count =
            ChatUnreadCounting.previewUnreadEstimate(
                lastMessageAt = conversation.lastMessageAt,
                lastMessageSenderUserId = ChatUnreadCounting.directPreviewSender(conversation),
                cursor = cursorStore.cursor(ChatReadThreadKey(ChatReadThreadKind.Direct, key)),
                currentUserId = currentUserId,
                trackingStartedAt = cursorStore.trackingStartedAt,
            )
        _unreadCountByConversationId.update { map ->
            if (count > 0) map + (key to count) else map - key
        }
    }

    fun recomputeAll(
        squadMessagesByCircleId: Map<String, List<CircleChatMessageDto>>,
        directMessagesByConversationId: Map<String, List<DirectMessageDto>>,
        directConversations: List<DirectConversationDto>,
    ) {
        val trackingStartedAt = cursorStore.trackingStartedAt
        val squads = mutableMapOf<String, Int>()
        squadMessagesByCircleId.forEach { (circleId, messages) ->
            val key = circleId.trim().takeIf { it.isNotEmpty() } ?: return@forEach
            val count =
                ChatUnreadCounting.squadUnreadCount(
                    messages = messages,
                    cursor = cursorStore.cursor(ChatReadThreadKey(ChatReadThreadKind.Squad, key)),
                    currentUserId = currentUserId,
                    trackingStartedAt = trackingStartedAt,
                )
            if (count > 0) squads[key] = count
        }

        val directs = mutableMapOf<String, Int>()
        directConversations.forEach { conversation ->
            val key = conversation.id.trim().takeIf { it.isNotEmpty() } ?: return@forEach
            val cached = directMessagesByConversationId[key].orEmpty()
            val count =
                if (cached.isEmpty()) {
                    ChatUnreadCounting.previewUnreadEstimate(
                        lastMessageAt = conversation.lastMessageAt,
                        lastMessageSenderUserId = ChatUnreadCounting.directPreviewSender(conversation),
                        cursor = cursorStore.cursor(ChatReadThreadKey(ChatReadThreadKind.Direct, key)),
                        currentUserId = currentUserId,
                        trackingStartedAt = trackingStartedAt,
                    )
                } else {
                    ChatUnreadCounting.directUnreadCount(
                        messages = cached,
                        cursor = cursorStore.cursor(ChatReadThreadKey(ChatReadThreadKind.Direct, key)),
                        currentUserId = currentUserId,
                        trackingStartedAt = trackingStartedAt,
                    )
                }
            if (count > 0) directs[key] = count
        }

        _unreadCountByCircleId.value = squads
        _unreadCountByConversationId.value = directs
    }

    fun maybeMarkSquadReadIfEligible(
        circleId: String,
        messages: List<CircleChatMessageDto>,
        isPinnedToBottom: Boolean = false,
        lastReadMessageId: String? = null,
    ) {
        markSquadReadIfChatTabVisible(circleId, messages)
    }

    fun maybeMarkDirectReadIfEligible(
        conversationId: String,
        messages: List<DirectMessageDto>,
        isPinnedToBottom: Boolean = false,
        lastReadMessageId: String? = null,
    ) {
        markDirectReadIfThreadVisible(conversationId, messages)
    }

    fun handleSquadMessageUpsert(
        circleId: String,
        messages: List<CircleChatMessageDto>,
        isPinnedToBottom: Boolean = false,
        lastReadMessageId: String? = null,
    ) {
        maybeMarkSquadReadIfEligible(circleId, messages)
        val key = circleId.trim()
        if (squadChatTabVisibleCircleId == key) return
        recomputeSquad(circleId, messages)
    }

    fun handleDirectMessageUpsert(
        conversationId: String,
        messages: List<DirectMessageDto>,
        isPinnedToBottom: Boolean = false,
        lastReadMessageId: String? = null,
    ) {
        maybeMarkDirectReadIfEligible(conversationId, messages)
        val key = conversationId.trim()
        if (directThreadVisibleConversationId == key) return
        recomputeDirect(conversationId, messages)
    }
}

File: otto-mobile/otto-mobile/ChatUnreadTracker.swift
import Combine
import Foundation

@MainActor
final class ChatUnreadTracker: ObservableObject {
    @Published private(set) var unreadCountByCircleID: [String: Int] = [:]
    @Published private(set) var unreadCountByConversationID: [String: Int] = [:]

    var totalChatUnreadCount: Int {
        unreadCountByCircleID.values.reduce(0, +) + unreadCountByConversationID.values.reduce(0, +)
    }

    private let cursorStore = ChatReadCursorStore()

    init() {}

    nonisolated deinit {}

    private var currentUserID = ""
    private(set) var squadChatTabVisibleCircleID: String?
    private(set) var directThreadVisibleConversationID: String?

    func markRead(thread: ChatReadThreadKey, messageId: String, at readAt: Date) {
        cursorStore.markRead(thread: thread, messageId: messageId, at: readAt)
    }

    func bind(currentUserID rawUserID: String) {
        let trimmed = rawUserID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed != currentUserID else { return }
        currentUserID = trimmed
        cursorStore.bind(userID: trimmed.isEmpty ? nil : trimmed)
    }

    func clearAll() {
        unreadCountByCircleID = [:]
        unreadCountByConversationID = [:]
        cursorStore.clearAll()
    }

    func setSquadChatTabVisible(circleID: String?, isVisible: Bool) {
        let key = circleID?.trimmingCharacters(in: .whitespacesAndNewlines)
        if isVisible, let key, !key.isEmpty {
            squadChatTabVisibleCircleID = key
        } else if !isVisible, squadChatTabVisibleCircleID == key || key == nil {
            squadChatTabVisibleCircleID = nil
        }
    }

    func isSquadChatTabVisible(circleID: String) -> Bool {
        let key = circleID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !key.isEmpty else { return false }
        return squadChatTabVisibleCircleID == key
    }

    func setDirectThreadVisible(conversationID: String?, isVisible: Bool) {
        let key = conversationID?.trimmingCharacters(in: .whitespacesAndNewlines)
        if isVisible, let key, !key.isEmpty {
            directThreadVisibleConversationID = key
        } else if !isVisible, directThreadVisibleConversationID == key || key == nil {
            directThreadVisibleConversationID = nil
        }
    }

    func unreadCount(forCircleID circleID: String) -> Int {
        let key = circleID.trimmingCharacters(in: .whitespacesAndNewlines)
        return unreadCountByCircleID[key] ?? 0
    }

    func unreadCount(forConversationID conversationID: String) -> Int {
        let key = conversationID.trimmingCharacters(in: .whitespacesAndNewlines)
        return unreadCountByConversationID[key] ?? 0
    }

    func markSquadReadIfChatTabVisible(circleID: String, messages: [CircleChatMessageDTO]) {
        let key = circleID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !key.isEmpty, squadChatTabVisibleCircleID == key else { return }
        guard let newest = messages.last else {
            unreadCountByCircleID.removeValue(forKey: key)
            return
        }
        cursorStore.markRead(thread: .squad(key), messageId: newest.id, at: newest.createdAt)
        unreadCountByCircleID.removeValue(forKey: key)
    }

    func markDirectReadIfThreadVisible(conversationID: String, messages: [DirectMessageDTO]) {
        let key = conversationID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !key.isEmpty, directThreadVisibleConversationID == key else { return }
        guard let newest = messages.last else {
            unreadCountByConversationID.removeValue(forKey: key)
            return
        }
        cursorStore.markRead(thread: .direct(key), messageId: newest.id, at: newest.createdAt)
        unreadCountByConversationID.removeValue(forKey: key)
    }

    func recomputeSquad(circleID: String, messages: [CircleChatMessageDTO]) {
        let key = circleID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !key.isEmpty else { return }
        let cursor = cursorStore.cursor(for: .squad(key))
        let count = ChatUnreadCounting.squadUnreadCount(
            messages: messages,
            cursor: cursor,
            currentUserID: currentUserID,
            trackingStartedAt: cursorStore.trackingStartedAt
        )
        if count > 0 {
            unreadCountByCircleID[key] = count
        } else {
            unreadCountByCircleID.removeValue(forKey: key)
        }
    }

    func recomputeDirect(conversationID: String, messages: [DirectMessageDTO]) {
        let key = conversationID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !key.isEmpty else { return }
        let cursor = cursorStore.cursor(for: .direct(key))
        let count = ChatUnreadCounting.directUnreadCount(
            messages: messages,
            cursor: cursor,
            currentUserID: currentUserID,
            trackingStartedAt: cursorStore.trackingStartedAt
        )
        if count > 0 {
            unreadCountByConversationID[key] = count
        } else {
            unreadCountByConversationID.removeValue(forKey: key)
        }
    }

    func recomputeDirectFromPreview(
        conversationID: String,
        lastMessageAt: Date?,
        lastMessageSenderUserID: String?
    ) {
        let key = conversationID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !key.isEmpty else { return }
        let cursor = cursorStore.cursor(for: .direct(key))
        let count = ChatUnreadCounting.previewUnreadEstimate(
            lastMessageAt: lastMessageAt,
            lastMessageSenderUserID: lastMessageSenderUserID,
            cursor: cursor,
            currentUserID: currentUserID,
            trackingStartedAt: cursorStore.trackingStartedAt
        )
        if count > 0 {
            unreadCountByConversationID[key] = count
        } else {
            unreadCountByConversationID.removeValue(forKey: key)
        }
    }

    func recomputeAll(
        squadMessagesByCircleID: [String: [CircleChatMessageDTO]],
        directMessagesByConversationID: [String: [DirectMessageDTO]],
        directConversations: [DirectConversationDTO]
    ) {
        let trackingStartedAt = cursorStore.trackingStartedAt
        var squads: [String: Int] = [:]
        for (circleID, messages) in squadMessagesByCircleID {
            let key = circleID.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !key.isEmpty else { continue }
            let count = ChatUnreadCounting.squadUnreadCount(
                messages: messages,
                cursor: cursorStore.cursor(for: .squad(key)),
                currentUserID: currentUserID,
                trackingStartedAt: trackingStartedAt
            )
            if count > 0 { squads[key] = count }
        }

        var directs: [String: Int] = [:]
        for conversation in directConversations {
            let conversationID = conversation.id.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !conversationID.isEmpty else { continue }
            let cached = directMessagesByConversationID[conversationID] ?? []
            if cached.isEmpty {
                let previewCount = ChatUnreadCounting.previewUnreadEstimate(
                    lastMessageAt: conversation.lastMessageAt,
                    lastMessageSenderUserID: conversation.lastMessage?.senderUserId,
                    cursor: cursorStore.cursor(for: .direct(conversationID)),
                    currentUserID: currentUserID,
                    trackingStartedAt: trackingStartedAt
                )
                if previewCount > 0 { directs[conversationID] = previewCount }
            } else {
                let count = ChatUnreadCounting.directUnreadCount(
                    messages: cached,
                    cursor: cursorStore.cursor(for: .direct(conversationID)),
                    currentUserID: currentUserID,
                    trackingStartedAt: trackingStartedAt
                )
                if count > 0 { directs[conversationID] = count }
            }
        }

        unreadCountByCircleID = squads
        unreadCountByConversationID = directs
    }

    func maybeMarkSquadReadIfEligible(
        circleID: String,
        messages: [CircleChatMessageDTO],
        isPinnedToBottom: Bool = false,
        lastReadMessageId: String? = nil
    ) {
        markSquadReadIfChatTabVisible(circleID: circleID, messages: messages)
    }

    func maybeMarkDirectReadIfEligible(
        conversationID: String,
        messages: [DirectMessageDTO],
        isPinnedToBottom: Bool = false,
        lastReadMessageId: String? = nil
    ) {
        markDirectReadIfThreadVisible(conversationID: conversationID, messages: messages)
    }

    func handleSquadMessageUpsert(
        circleID: String,
        messages: [CircleChatMessageDTO],
        isPinnedToBottom: Bool = false,
        lastReadMessageId: String? = nil
    ) {
        maybeMarkSquadReadIfEligible(circleID: circleID, messages: messages)
        let key = circleID.trimmingCharacters(in: .whitespacesAndNewlines)
        if squadChatTabVisibleCircleID == key { return }
        recomputeSquad(circleID: circleID, messages: messages)
    }

    func handleDirectMessageUpsert(
        conversationID: String,
        messages: [DirectMessageDTO],
        isPinnedToBottom: Bool = false,
        lastReadMessageId: String? = nil
    ) {
        maybeMarkDirectReadIfEligible(conversationID: conversationID, messages: messages)
        let key = conversationID.trimmingCharacters(in: .whitespacesAndNewlines)
        if directThreadVisibleConversationID == key { return }
        recomputeDirect(conversationID: conversationID, messages: messages)
    }
}
